### PR TITLE
jitter: add enable-jitter-dynamic configure option

### PR DESCRIPTION
**Pull request by pmonreal**
Add a new configure option called enable-jitter-dynamic to allow building with jitterentropy using the dynamic libjitterentropy.so library instead of only allowing the static libjitterentropy.a.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Why would our build scripts care about static/dynamic linking of libjitterentropy?  I'd have thought accepting whatever was installed was preferable.

Can anyone think of a use case for the distinction?

**Comment by pmonreal** [comment]
Right, at the moment only the static libjitterentropy.a can be used as its hard-coded in [here](https://github.com/openssl/openssl/blob/241d4826f8ee39c92d9b3233146c3e12314871ec/Configurations/00-base-templates.conf#L110). This change adds a configuration option to also allow using the dynamic libjitterentropy.so library following the convention used in `enable-{brotli,zlib,zstd}-dynamic`. Regarding FIPS validations, I think there is no problem to obtain an ESV entropy assessment and validation from NIST.

**Comment by paulidale** [comment]
The easier fix would be to switch that line to not force the .a file.  Adding configuration options and conditional code seems like overkill here.

I've not tried to validate jitterentropy so don't know for sure.  I've heard that it's possible.

**Comment by pmonreal** [comment]
Yes, I would also prefer just using plain `-ljitterentropy`. In that case `libjitterentropy.so` would take compiler preference over `libjitterentropy.a`. Maybe @xnox has some comments as the designer of this in #24844.

**Comment by xnox** [comment]
There are two types of linking that is supported right now:
- inside the FIPS module
- outside of the FIPS module

And the symbols leak.

If one certifies jitterentropy it can be difficult to guarantee that the same one is is used with the same module as submitted.

Also, I wanted to preserve the same linker flags, as lots of software breaks linking if libcrypto.so now needs to link with libjitterentropy.

I haven't completed the work, but also wanted to merge libcrypto.a and libjitterentrooy.a such that for static linking too the build flags are the same.

If you are doing new submissions it is best to statically link jitterentropy into fips.so. But do note there are bugs currently that cause certain functionality evade that entropy source inside FIPS.so and escape.

Plus if you want to use the same certified build of jitterentropy in libgcrypt, BouncyCastle, aws-lc, you way want to use static linking throughout.

Similarly it is easier to explain secure channel with statically linked binaries.

For https://csrc.nist.gov/projects/cryptographic-module-validation-program/entropy-validations/certificate/191 from public use document you can see that I only validated static libjitterentropy.a and use it throughout all builds of libcrypto and keep it outside the FIPS cryptographic module boundary.

If you do decide to lift this static linking restriction, ensure to do the following:
- check that generated pkg config file is desirable
- build all reverse dependencies and ensure it all works
- check that any toolchains that dlopen libcrypto alone work as expected (dotnet, go openssl, etc)
- ensure your lab is happy to certify dynamically linked entropy source
- when submitting entropy source validation, do certify both static and dynamic builds just in case
- check you can use certified build everywhere you wish to reuse the same entropy source

So far I have managed to reuse statically linked entropy source for over 1.5 years now across multiple toolchain transitions. I don't know if dynamically linked one would work just as good.

Do consider that current entropy sources do not have an expiry date, and often can outlive multiple cryptographic module submissions - if validated stand alone and operated outside of the module boundary.

For a while I also had newer builds of shared jitterentropy library installed without affecting libcrypto.

So for me this was done out of abundance of caution.

Bouncy castle jitter uses a patched dynamically linked jitterentropy. Libgcrypt uses staticly linked one. Kernel does statically linked one into vmlinuz. Aws-lc started off with statically linked one, but I cannot quite tell how dynamically linked one looks like - please check the 3.0 FIPS branch that is in progress, as it has separate dynamic and static builds.

For me static linking looked like the easiest path to have everything working correctly for a shared libcrypto.so + fips.so, with two submissions in process that use this design, which also works with openssl 3.0.9 and 3.1.2 existing modules. If at all possible, I would want to preserve this statically linked option - even when dynamically linked jitterentropy is available (or I would be forced to spend money to re-validate dynamic jitterentropy which i didn't validate 😞). As such design does not cause leakage of libjitter for dynamically linked users of OpenSSL.

It is a gap to merge static jitterentropy into static libcrypto.a, as for jitterentropy still leaks for private linking. But possibly it is good that that leaks, as do note the other caveat that there needs to be a unique FIPS.so for each unique libcrypto that tries to load it, hence I try to prevent apps from statically linking libcrypto as much as possible.

So yeah, above choice of static linking was to simplify entropy validation, FIPS submission, and to be a perfect mimic of existing linking of just `-lcrypto -lssl` for consumers of OpenSSL to reduce widespread FTBFS of reverse dependencies. And left here as a word of caution that alternative path of shared linking is slightly unchartered territory.

Make your own choices, talk to your lab, and yeah as long as there is a way to choose type of linking of this library uses cases will remain flexible.

I am also curious about nonstop platform and if they are interested in this or not.

It would be nice if we could embed HMAC inside FIPS.so upstream like many distributions do. And make it possible to statically link libcrypto + libjitter + fips whilst having boundary around fips portion alone, like boringcrypto and aws-lc support as it would be very liberating to have tight FIPS boundary, reusable entropy source, and yet support static linking for binaries that need relocation / redistribution.

Note I wonder if for non-fips cases it makes sense to always use linked jitterentropy anyway. Performance is great, especially if compiled with oversampling rate 3, kernel uses the same anyway or uses lower oversampling rate, and even with vdso it seems like there is performance penalty.

Oh, I didn't benchmark performance with shared linking either.

So yeah, that was my thinking, and at the time I was all alone and nobody to really discuss this with, and I didn't have resources to safely land dynamically linked jitterentropy.

I do not believe there is anything inherently impossible with using dynamic linking for new submissions. Just likely lots of additional work that was never been done before at large scale in a typical Linux distribution.

Do note that jitterentropy does evolve slowly and can add or change symbols from time to time too. Thus for me, I chose to use jitter implementation of sha3 for conditioning function, such that openssl one is not used, to ensure the entropy source is standalone and reusable across different cryptographic modules (openssl and non-openssl ones).

I don't know when, but it seems like there are slow efforts to standardize XDRNG based on XOF Shake, that would help with performance even more once that is available.

Hope this helps.

**Review by xnox — COMMENTED** [review]
Reviewing the patches, I like this. It is worth preserving explicit choice of link type.

Note, there are different requirements in behaviour when entropy source is inside the cryptographic module boundary, versus outside of it.

LGTM to land it as a new option.

**Comment by t8m** [comment]
Could you please add `enable-jitter-dynamic` temporarily to the run-checker-ci.yml (in a separate commit with DROP annotation in the commit title) and also with a regular commit to run-checker-daily.yml ?

**Comment by pmonreal** [comment]
Thanks @xnox for your feedback. I think preserving the choice explicitly can help when choosing static or dynamic linking. We have been using dynamic linking for a couple of years at SUSE/openSUSE and we are haven't found any issue. Also, our ESV certificate was already done for the .so lib.

@t8m I have added two commits with the requested CI changes. Please, let me know if that's fine. TIA

**Comment by xnox** [comment]
I like that both options will be in active use in production.

**Review by t8m — DISMISSED** [review]

**Comment by pmonreal** [comment]
I just updated the PR to resolve conflicts with master and fixed the `enable-jitter-dynamic` test in `.github/workflows/run-checker-daily.yml`. Should be ready now.

**Comment by pmonreal** [comment]
Hello, @t8m! The requested changes for `run-checker-ci.yml` will fail with errors like `symbols jent_entropy_* not in the allowed platform symbols list` when running the `checkplatformsyms.pl` script and it would require to include the jitterentropy symbols in `util/platform_symbols/*-symbols.txt`. Note that, the symbols check will fail for both the `enable-jitter` and `enable-jitter-dynamic` configure options. These options also require additional configuration that cannot be easily fine-grained in this workflow. Could we skip the `run-checker-ci.yml` part as it just complicates things and it would fail for both the already accepted `enable-jitter` option and also for the new `enable-jitter-dynamic`? I guess the `run-checker-daily.yml` should be enough. TIA.

**Comment by xnox** [comment]
> Hello, @t8m! The requested changes for `run-checker-ci.yml` will fail with errors like `symbols jent_entropy_* not in the allowed platform symbols list` when running the `checkplatformsyms.pl` script and it would require to include the jitterentropy symbols in `util/platform_symbols/*-symbols.txt`. Note that, the symbols check will fail for both the `enable-jitter` and `enable-jitter-dynamic` configure options. These options also require additional configuration that cannot be easily fine-grained in this workflow. Could we skip the `run-checker-ci.yml` part as it just complicates things and it would fail for both the already accepted `enable-jitter` option and also for the new `enable-jitter-dynamic`? I guess the `run-checker-daily.yml` should be enough. TIA.


Should these symbols be allowed, or like not exported?

File: doc/man7/EVP_RAND-JITTER.pod [hunk 1]
```diff
@@ -13,8 +13,8 @@ This software seed source produces randomness based on tiny CPU
 "jitter" fluctuations.
 
 It is available when OpenSSL is compiled with B<enable-jitter>
-option. When available it is listed in B<openssl list
--random-generators> and B<openssl info -seeds>.
+option or B<enable-jitter-dynamic>. When available it is listed in
+B<openssl list -random-generators> and B<openssl info -seeds>.
```
Comment (paulidale) [inline doc/man7/EVP_RAND-JITTER.pod]: You should briefly mention the difference between the two options here.  It only needs to be a one-liner:
"The first statically links the library while the second dynamically links it."

Comment (pmonreal) [reply doc/man7/EVP_RAND-JITTER.pod]: Thanks for the comment! I have updated the PR. See also the comment from @xnox in [here](https://github.com/openssl/openssl/pull/29064#issuecomment-3911271594).

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Nice work.

**Comment by pmonreal** [comment]
Hi, @quarckster @t8m! Could this be considered for OpenSSL 4.0.0? TIA

**Comment by Mulling** [comment]
I second Pedro, it would be nice for us (SUSE) if this could be included in 4.0.0.

File: .github/workflows/run-checker-ci.yml [hunk 2]
```diff
@@ -41,18 +41,33 @@ jobs:
...
           no-ui,
           no-quic
         ]
     runs-on: ubuntu-latest
     steps:
+    - name: checkout jitter
+      uses: actions/checkout@v5
+      with:
+        repository: smuellerDD/jitterentropy-library
+        ref: v3.6.3
```
Comment (esyr) [inline .github/workflows/run-checker-ci.yml:54]: SHA is preferred over tag here, as the latter can be changed.

File: .github/workflows/run-checker-ci.yml [hunk 3]
```diff
@@ -41,18 +41,33 @@ jobs:
...
+        path: jitter
+        persist-credentials: false
+    - name: build jitter
+      run: |
+        pushd jitter/
+        make
+        ln -s libjitterentropy.so.3.6.3 libjitterentropy.so
+        ln -s libjitterentropy.so.3.6.3 libjitterentropy.so.3
```
Comment (esyr) [inline .github/workflows/run-checker-ci.yml:62]: `make install-shared` can help avoiding these commands.

**Comment by esyr** [comment]
> Could this be considered for OpenSSL 4.0.0?

I don't think so, it's a feature, and we are past feature freeze.

**Comment by xnox** [comment]
@pmonreal could you please double check with your lab w.r.t. claims you can make when entropy source is dynamically linked? There have been some additional clarifications and requirements from NIST.

**Comment by pmonreal** [comment]
> @pmonreal could you please double check with your lab w.r.t. claims you can make when entropy source is dynamically linked? There have been some additional clarifications and requirements from NIST.

I just read the new entropy requirements and I have asked our lab.

**Comment by pmonreal** [comment]
> > @pmonreal could you please double check with your lab w.r.t. claims you can make when entropy source is dynamically linked? There have been some additional clarifications and requirements from NIST.
> 
> I just read the new entropy requirements and I have asked our lab.

Hi, @xnox! Our FIPS lab doesn't see any new NIST claims that might affect using jitterentropy dynamically, note that we have used this approach in the past certification rounds with no issues. Do you see any specific new requirement that should be clarified? For FIPS validation using dynamic linking you would need an ESV certificate and the library must be integrity tested as any software components within the boundary. I will update the PR to address the comments from @esyr asap.

**Comment by andrewkdinh** [comment]
ping @pmonreal

**Review by paulidale — CHANGES_REQUESTED** [review]
An AI review of this:

1. Dynamic linkage breaks jitterentropy state isolation. Each OpenSSL JITTER instantiation calls `jent_entropy_init_ex()`, while its locking hooks are no-ops ([source](https://github.com/openssl/openssl/blob/2be8776aa174d45b64a7bdb94851b3ff3cccb03d/providers/implementations/rands/seed_src_jitter.c#L147-L165)). With the new shared-library linkage, `libcrypto`, `fips.so`, and unrelated consumers share one process-global jitterentropy state. Even v3.7.0—newer than the pinned v3.6.3—states initialization is not thread-safe and must occur exactly once before concurrent use ([upstream contract](https://github.com/smuellerDD/jitterentropy-library/blob/v3.7.0/src/jitterentropy-base.c#L955-L960)). Upstream subsequently reproduced five races under TSAN and fixed the global state and callback publication in [8e12074](https://github.com/smuellerDD/jitterentropy-library/commit/8e12074eb6364e33fe4bdfec59798b7258d2a22f) and [c5bf563](https://github.com/smuellerDD/jitterentropy-library/commit/c5bf5635784ddcd56cb1c015bc47bc75e7bd0552), neither currently released. Static linkage already has same-image multi-libctx concerns, but this PR newly merges formerly independent libcrypto/FIPS/foreign-consumer state. Dynamic mode should require a released upstream fix or retain process isolation.
2. The new CI does not perform a valid dynamic-jitter test. Both jobs check out and build jitter first, then check out OpenSSL at the workspace root ([daily workflow](https://github.com/openssl/openssl/blob/2be8776aa174d45b64a7bdb94851b3ff3cccb03d/.github/workflows/run-checker-daily.yml#L201-L226)). Because the root has no matching `.git`, `actions/checkout` marks it for removal and deletes its contents ([checkout implementation](https://github.com/actions/checkout/blob/v5/src/git-directory-helper.ts#L27-L31), [deletion](https://github.com/actions/checkout/blob/v5/src/git-directory-helper.ts#L117-L123)), including `jitter/`. Reverse the checkout order.
  After that fix, `opensslwrap version -c` still lacks `LD_LIBRARY_PATH=jitter`; in the DROP job, that assignment appears only in the step’s display name while `run:` remains plain `make test` ([lines 75–82](https://github.com/openssl/openssl/blob/2be8776aa174d45b64a7bdb94851b3ff3cccb03d/.github/workflows/run-checker-ci.yml#L75-L82)). Set the loader path job-wide and verify the resolved SONAME so CI cannot silently use a system copy.
3. The shared dependency is applied to unrelated artifacts, including non-jitter FIPS. The new flag is added to global `ex_libs` ([source](https://github.com/openssl/openssl/blob/2be8776aa174d45b64a7bdb94851b3ff3cccb03d/Configurations/00-base-templates.conf#L107-L113)). A dry run of `enable-jitter-dynamic enable-fips` without `enable-fips-jitter` still links `providers/fips.so` with `-l:libjitterentropy.so`. On linkers without `--as-needed`, that creates an unused `DT_NEEDED` dependency—the normal GNU ld behavior ([documentation](https://sourceware.org/binutils/docs/ld/Options.html#Options)). The previous archive did not have this behavior because unused members were not extracted. Scope the dependency to artifacts containing JITTER references.
4. The linker spelling is not portable across targets that accept the option. `BASE_unix` emits GNU-specific `-l:libjitterentropy.so`. darwin64-arm64 configures successfully but Apple ld interprets this as library :libjitterentropy.so and fails; Darwin’s shared artifact is a `.dylib`. Use target-aware naming/link semantics or explicitly gate and document the supported platforms.
5. The FIPS boundary documentation becomes misleading. The documentation still calls this an “internal jitter [source](https://github.com/openssl/openssl/blob/2be8776aa174d45b64a7bdb94851b3ff3cccb03d/doc/man7/EVP_RAND-JITTER.pod#L52-L59)” (source), but with dynamic linkage the entropy implementation is an external DSO, while OpenSSL’s module-integrity POST authenticates only `fips.so`. The exact DSO identity, boundary placement, loading controls, and integrity responsibility need to be stated explicitly for `enable-fips-jitter enable-jitter-dynamic`.
6. Configuration state is inconsistent. Unlike zlib/brotli/zstd, `jitter` does not cascade-disable `jitter-dynamic`. `./Configure enable-jitter-dynamic no-jitter` reports `jitter-dynamic` enabled while defining `OPENSSL_NO_JITTER`. Add `"jitter" => ["jitter-dynamic"]` to the cascades and test option ordering.

I've **not** going through and manually checked each of these items but some do look relevant.

It also looks like a rebase is required.
